WIP: Allow module extensions to persist facts in the lockfile

fmeum opened the pull request.

fmeum commented:
I will add freeform docs after the first review round.

fmeum commented:
@bazel-io fork 8.4.0

meteorcloudy commented on src/main/java/com/google/devtools/build/lib/bazel/bzlmod/SingleExtensionEvalFunction.java:
```diff
@@ -229,7 +233,20 @@ public SkyValue compute(SkyKey skyKey, Environment env)
+      throw new SingleExtensionEvalFunctionException(
+          ExternalDepsException.withMessage(
+              Code.BAD_LOCKFILE,
+              "The module extension '%s' has changed its facts: %s != %s",
```
Should we also recommend what the solution could be?

meteorcloudy reviewed (commented):
Looks great! But please wait also for @Wyverald 's review.

I have only one question about when old facts are going to be deleted so that the lockfile won't grow forever. IIUC, the module extension will always pass new facts via `extension_metadata` which should refresh the lockfile, right?

fmeum commented:
@Wyverald Friendly ping

Wyverald commented on src/main/java/com/google/devtools/build/lib/bazel/bzlmod/FactsAdapter.java:
```diff
@@ -0,0 +1,71 @@
+  public void write(JsonWriter out, Facts facts) throws IOException {
+    String json;
+    try {
+      json = Json.INSTANCE.encode(facts.value());
```
just curious -- how come we don't need the `INTERNAL_BAZEL_ONLY_UTF_8_BYTE_STRINGS` thing here?

fmeum replied:
`encode` doesn't emit Unicode escape sequences for non-ASCII Unicode characters, it just dumps out their bytes verbatim - which happens to preserve raw UTF-8 bytes.

Wyverald commented on src/test/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleExtensionResolutionTest.java:
```diff
@@ -2528,6 +2537,116 @@ public void extensionMetadata_isolatedDev() throws Exception {
+    var result =
+        evaluateSimpleModuleExtension(
+            """
+            return ctx.extension_metadata(\
```
why all these backslashes?

fmeum replied:
The function requires the arg to be a single line or the indentation will be off.

fmeum replied:
Didn't know this exists until now, updated

fmeum commented on src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BazelLockFileValue.java:
```diff
@@ -131,6 +132,8 @@ static Builder builder() {
           ModuleExtensionId, ImmutableMap<ModuleExtensionEvalFactors, LockFileModuleExtension>>
       getModuleExtensions();
 
+  public abstract ImmutableMap<ModuleExtensionId, Facts> getFacts();
```
That would affect the JSON structure. My first attempt was to merge the facts in as a special "factor", but that turned out to be too hacky. I'm not a fan of introducing yet another nesting level to distinguish between facts and locked extension results as this just makes diffs more ugly. Flat structures merge better :-)

fmeum commented:
> This is really just arbitrary data that an extension can store for later use.

What I intend to get across with `facts` is that the data should be such that it can't get stale (only unused) over time - not fully arbitrary.

dzbarsky commented:
I am trying to use this for Rust rules and running into the max-depth limitation. My facts are structured as follows:
```
facts = {
    "{crate_name}_{version}": {
        "dependencies": { crates.io dependencies API response}
        "features": { crates.io crate API response}
    },
}
``` 

Those API responses themselves have a bit of nesting. So here's a materialized example:
```
{
 "zvariant_utils_1.0.1": {
    "dependencies": [
      {
        "crate_id": "syn",
        "default_features": true,
        "downloads": 0,
        "features": [
          "extra-traits",
          "full"
        ],
        "id": 6590504,
        "kind": "normal",
        "optional": false,
        "req": "^1.0.103",
        "target": null,
        "version_id": 800172
      }
    ],
    "features": {}
  }
}
```

```
➜  smoke git:(main) ✗ jq '                                                        
  [paths as $p | {path:$p, depth:($p|length)}]
  | max_by(.depth)
  | .path
' /tmp/facts.json
[
  "zvariant_utils_1.0.1",
  "dependencies",
  0,
  "features",
  1
]
```

It would be nice to allow an extra 1-2 levels of nesting to cover this sort of use case. (currently it seems only 4 are supported, despite the error saying 5)

fmeum commented:
@dzbarsky and I discussed possible restrictions that would make facts work better for end users. I'll update the PR with a new commit soon.

aignas commented:
+1 about nesting.

I would like to use them for `rules_python` for the following:
* distribution name to list of URLs mapping by index URL.
* distribution name to list of deps by version of the wheel.

So the schema might be something like:
```
{
  "pydantic": {
    "srcs": {
      "pypi.org/simple": [
        "one",
        "different",
      ],
    "deps": {
      "1.1.0": [
        "foo; python_version > 3.11",
      ],
    },
  }
}
```

fmeum commented:
@Wyverald I implemented some of the ideas I developed with @dzbarsky. Extensions will be able to benefit from `go.sum` style merge conflict resolution. `mctx.facts` is restricted to only support `[...]`, `.get(...)` and `in`, not enumeration. This is modeled after how providers work and pushed implementations towards disregarding additional keys, which should make any kind of merge conflict resolution more reliable.

Mid to long term we could consider moving the merge drive into the Bazel client (not the server, as that could cause server restarts during git operations). For now, I just updated the jq script, but users would have to update it in their git config to benefit from the changes.

thesayyn commented:
It would be nice to land this api into Bazel 8.

fmeum commented:
@Wyverald Addressed the comments and also wrote some long-form docs
